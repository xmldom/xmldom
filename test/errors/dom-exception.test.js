'use strict';
const { describe, expect, test } = require('@jest/globals');
const { DOMException, DOMExceptionName } = require('../../lib/errors');

describe('DOMException', () => {
	const CODES = [
		DOMException.INDEX_SIZE_ERR,
		DOMException.DOMSTRING_SIZE_ERR,
		DOMException.HIERARCHY_REQUEST_ERR,
		DOMException.WRONG_DOCUMENT_ERR,
		DOMException.INVALID_CHARACTER_ERR,
		DOMException.NO_DATA_ALLOWED_ERR,
		DOMException.NO_MODIFICATION_ALLOWED_ERR,
		DOMException.NOT_FOUND_ERR,
		DOMException.NOT_SUPPORTED_ERR,
		DOMException.INUSE_ATTRIBUTE_ERR,
		DOMException.INVALID_STATE_ERR,
		DOMException.SYNTAX_ERR,
		DOMException.INVALID_MODIFICATION_ERR,
		DOMException.NAMESPACE_ERR,
		DOMException.INVALID_ACCESS_ERR,
		DOMException.VALIDATION_ERR,
		DOMException.TYPE_MISMATCH_ERR,
		DOMException.SECURITY_ERR,
		DOMException.NETWORK_ERR,
		DOMException.ABORT_ERR,
		DOMException.URL_MISMATCH_ERR,
		DOMException.QUOTA_EXCEEDED_ERR,
		DOMException.TIMEOUT_ERR,
		DOMException.INVALID_NODE_TYPE_ERR,
		DOMException.DATA_CLONE_ERR,
	];

	test.each(CODES)('should have correct name from code %s', (code) => {
		const name = new DOMException(code).name;
		expect(name).not.toBeUndefined();
		expect(DOMExceptionName).toHaveProperty(name);
		expect(new DOMException('test', name).code).toBe(code);
		expect(name).toMatchSnapshot();
	});
	test.each([-1, 0, 26])('should use invalid code %s as message', (code) => {
		const err = new DOMException(code);
		expect(err.message).toBe(code);
		expect(err.name).toBe(DOMExceptionName.Error);
		expect(err.code).toBe(0);
	});
	test('should be an instance of Error', () => {
		expect(new DOMException('') instanceof Error).toBe(true);
	});

	test('should be an instance of DOMException', () => {
		expect(new DOMException('') instanceof DOMException).toBe(true);
	});

	test('should store first string argument as message', () => {
		const error = new DOMException('FROM TEST');
		expect(error.message).toBe('FROM TEST');
	});

	test('should have correct StackTrace', () => {
		const message = 'MESSAGE';
		const error = new DOMException(message, DOMExceptionName.SecurityError);
		const stack = error.stack && error.stack.split(/[\n\r]+/);
		expect(stack && stack.length).toBeGreaterThanOrEqual(2);
		expect(stack[0]).toBe(`${DOMExceptionName.SecurityError}: ${message}`);
		expect(stack[1]).toContain(__filename);
	});

	test('Error should not be instanceof DOMException', () => {
		expect(new Error() instanceof DOMException).toBe(false);
	});
});
