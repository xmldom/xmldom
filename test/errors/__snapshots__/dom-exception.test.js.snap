// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOMException should have correct name from code 1 1`] = `"IndexSizeError"`;

exports[`DOMException should have correct name from code 2 1`] = `"DomstringSizeError"`;

exports[`DOMException should have correct name from code 3 1`] = `"HierarchyRequestError"`;

exports[`DOMException should have correct name from code 4 1`] = `"WrongDocumentError"`;

exports[`DOMException should have correct name from code 5 1`] = `"InvalidCharacterError"`;

exports[`DOMException should have correct name from code 6 1`] = `"NoDataAllowedError"`;

exports[`DOMException should have correct name from code 7 1`] = `"NoModificationAllowedError"`;

exports[`DOMException should have correct name from code 8 1`] = `"NotFoundError"`;

exports[`DOMException should have correct name from code 9 1`] = `"NotSupportedError"`;

exports[`DOMException should have correct name from code 10 1`] = `"InUseAttributeError"`;

exports[`DOMException should have correct name from code 11 1`] = `"InvalidStateError"`;

exports[`DOMException should have correct name from code 12 1`] = `"SyntaxError"`;

exports[`DOMException should have correct name from code 13 1`] = `"InvalidModificationError"`;

exports[`DOMException should have correct name from code 14 1`] = `"NamespaceError"`;

exports[`DOMException should have correct name from code 15 1`] = `"InvalidAccessError"`;

exports[`DOMException should have correct name from code 16 1`] = `"ValidationError"`;

exports[`DOMException should have correct name from code 17 1`] = `"TypeMismatchError"`;

exports[`DOMException should have correct name from code 18 1`] = `"SecurityError"`;

exports[`DOMException should have correct name from code 19 1`] = `"NetworkError"`;

exports[`DOMException should have correct name from code 20 1`] = `"AbortError"`;

exports[`DOMException should have correct name from code 21 1`] = `"URLMismatchError"`;

exports[`DOMException should have correct name from code 22 1`] = `"QuotaExceededError"`;

exports[`DOMException should have correct name from code 23 1`] = `"TimeoutError"`;

exports[`DOMException should have correct name from code 24 1`] = `"InvalidNodeTypeError"`;

exports[`DOMException should have correct name from code 25 1`] = `"DataCloneError"`;
